consensus/istanbul/core/backlog.go

```diff
@@ -95,115 +108,178 @@ func (c *core) storeBacklog(msg *istanbul.Message, src istanbul.Validator) {
+// Return slice of sequences present in backlog sorted in ascending order
+// Call with backlogsMu held.
+func (c *core) getSortedBacklogSeqs() []uint64 {
+	backlogSeqs := make([]uint64, len(c.backlogBySeq))
+	i := 0
+	for k := range c.backlogBySeq {
+		backlogSeqs[i] = k
+		i++
+	}
+	sort.Slice(backlogSeqs, func(i, j int) bool {
+		return backlogSeqs[i] < backlogSeqs[j]
+	})
+	return backlogSeqs
+}
+
+// Drain a backlog for a given sequence, passing each to optional callback.
+// Call with backlogsMu held.
+func (c *core) drainBacklogForSeq(seq uint64, cb func(*istanbul.Message, istanbul.Validator)) {
+	backlogForSeq := c.backlogBySeq[seq]
+	if backlogForSeq == nil {
+		return
+	}
+
+	for !backlogForSeq.Empty() {
+		m := backlogForSeq.PopItem()
+		msg := m.(*istanbul.Message)
+		if cb != nil {
+			_, src := c.valSet.GetByAddress(msg.Address)
+			if src != nil {
+				cb(msg, src)
+			}
+		}
+		c.backlogCountByVal[msg.Address]--
+		c.backlogTotal--
+	}
+	delete(c.backlogBySeq, seq)
 }
 
 func (c *core) processBacklog() {
+
 	c.backlogsMu.Lock()
 	defer c.backlogsMu.Unlock()
 
-	for src, backlog := range c.backlogs {
-		if backlog == nil {
-			continue
-		}
+	for _, seq := range c.getSortedBacklogSeqs() {
 
 		logger := c.newLogger("func", "processBacklog", "from", src)
-		isFuture := false
-
-		// We stop processing if
-		//   1. backlog is empty
-		//   2. The first message in queue is a future message
-		for !(backlog.Empty() || isFuture) {
-			m, prio := backlog.Pop()
-			msg := m.(*istanbul.Message)
-			var view *istanbul.View
-			switch msg.Code {
-			case istanbul.MsgPreprepare:
-				var m *istanbul.Preprepare
-				err := msg.Decode(&m)
-				if err == nil {
-					view = m.View
+
+		if seq < c.currentView().Sequence.Uint64() {
+			// Earlier sequence. Prune all messages.
+			c.drainBacklogForSeq(seq, nil)
+		} else if seq == c.currentView().Sequence.Uint64() {
+			// Current sequence. Process all in order.
+			c.drainBacklogForSeq(seq, func(msg *istanbul.Message, src istanbul.Validator) {
+				var view *istanbul.View
+				switch msg.Code {
+				case istanbul.MsgPreprepare:
+					var m *istanbul.Preprepare
+					err := msg.Decode(&m)
+					if err == nil {
+						view = m.View
+					}
+				case istanbul.MsgPrepare:
+					fallthrough
+				case istanbul.MsgCommit:
+					var sub *istanbul.Subject
+					err := msg.Decode(&sub)
+					if err == nil {
+						view = sub.View
+					}
+				case istanbul.MsgRoundChange:
+					var rc *istanbul.RoundChange
+					err := msg.Decode(&rc)
+					if err == nil {
+						view = rc.View
+					}
 				}
-			case istanbul.MsgPrepare:
-				fallthrough
-			case istanbul.MsgCommit:
-				var sub *istanbul.Subject
-				err := msg.Decode(&sub)
-				if err == nil {
-					view = sub.View
+				if view == nil {
+					logger.Error("Nil view", "msg", msg)
+					// continue
+					return
 				}
-			case istanbul.MsgRoundChange:
-				var rc *istanbul.RoundChange
-				err := msg.Decode(&rc)
+				err := c.checkMessage(msg.Code, view)
```

Comment:
This feels buggy

Reply:
what specifically?

Reply:
I don't remember TBH

Reply:
Looking at this, I think we can evade signature checking here still. When a new message comes in, `handleMsg` checks signature + validator in current valset. Then it calls `handleCheckedMsg` which is where for future messages `storeBacklog` is called. By the time this callback handles the message again, the epoch could have changed, so the validator that sent it may no longer be in the active set. WDYT?